### ISSUE TITLE
Replace stutter, Pstutter, etc. with non-ableist alternatives

### ISSUE THREAD

**Issue by mossheim** [supercollider/supercollider#5383]
## Motivation

Per discussion at https://scsynth.org/t/terminology-question/3372, using "stutter" this way is ableist and actually also not a good description of what these features even do. We should replace them with clearer and more inclusive alternatives.

## Description of Proposed Feature

- Deprecate these functions and classes and replace them with new ones:
  - Dstutter
  - Pstutter
  - `Array:-stutter`
  - `List:-stutter`

## Plan for Implementation

Info on deprecation is here: https://github.com/supercollider/supercollider/wiki/git-workflow-and-guidelines#deprecation. Currently-deprecated things [here](https://github.com/supercollider/supercollider/tree/develop/SCClassLibrary/deprecated) may serve as a good guide/inspiration. Deprecating the actual UGen might prove difficult, this probably needs more discussion and can maybe be split into a separate issue if needed? The rest should be easy.

**Comment by brianlheim** [comment]
Other comments from that thread which are relevant to this discussion:

From @Sciss:

> There are three occurrences: `.stutter` on collections, `Pstutter` (patterns), and `Dstutter` (demand UGen).
> 
> There is this equivalence 
> ```supercollider
> [4, 5, 6].stutter(3)
> [4, 5, 6].collect(_.dup(3)).flatten
> [4, 5, 6].dup(3).flop.flatten
> ```
> The second "bracket" is due to SC lacking a `flatMap` operator (I guess it would be called `flatCollect`). Putting these terms together is probably not feasible. I would also not use terms like 'multiply', 'repeat', or 'times' that could be confused with arithmetic operations, or operations on the entire collection. Somehow the operation _on each element_ must be conveyed. And there is no `Each` tradition in SuperCollider (AFAIK), so there is no precedence of saying `dupEach` or `repeatEach`.
> 
> I guess `reproduce` could work, even though it's another metaphor, given that we imagine a bottom-up multiplication on the basis of individual elements of a population? In terms of a non-interpolating expansion, perhaps `stretch` could work as well, if it wasn't already used by patterns. Maybe `dilate`? Or if you think of a volume (balloon), perhaps `inflate`.
> 
> __Edit:__ Or if staying with the idea of getting stuck: `hang` (taken by `Condition`), `stick` (bit unclear), `stall`.

From gurk (not sure if this is the same person as www.github.com/gurk):

> Just had a thought while reading, may as well share.
> 
> How about reiterate instead of stutter?
> 
> Not entirely convinced of my own suggestion though.
> 
> **Edit:** also thinking of recur or reoccur, but that's probably a bit confusing.

**Comment by Sciss** [comment]
It's probably not wise to rename the server plugin without a backward compatible automatic fallback. But since the name in the class library (so what's user-facing) can be different, maybe the sensible solution would be to change it there and refer to the old UGen name internally?

**Comment by smrg-lm** [comment]
Tremolo (musical but not exact) o repeat (numpy has a function that does exactly that) could be considered.

The name should reflect the action based on some common vocabulary otherwise it will be misleading.

**Comment by smrg-lm** [comment]
`repeatEach` sounds good to me and it goes well with the already existing repeat method, `reproduce` and others doesn't clearly reflect the action for me. The name should reflect the action based on some common vocabulary otherwise it will be misleading. BTW, stutter was booth ableist and misleading.

**Comment by Sciss** [comment]
I was thinking of something like _tremolo_, but this doesn't necessarily make sense as a generic array operation which may be used on data for which a "tremolo" makes no sense; nor is "tremolo" defined precisely. I think we can approach this from two perspectives:

- 1. how easy is it for somebody new to pick up the meaning?
- 2. how easy is it to remember the function?

These are some factors:

- meaning can be instantaneously derived from the term. So `repeatEach` is probably understandable directly by most people.
- meaning can be derived from metaphor: `stutter` -> within a spoken utterance, repeat syllables. is often imprecise, for example here to concept of strict and constant repetition is not captured, so `stutter` is actually not even a good description.
- meaning can be imported from prior knowledge. What would the same or similar operations be called in other programming languages or frameworks? So if numpy has `repeat`, that would be plus point for using it. On the other hand, it collides with SuperCollider's missing distinction in nomenclature between methods that operate on a per-element basis and those that operate on the entire sequence. `dup` is an example of operating on the sequence, `*` is an example of operating on each element. There is no way to know this other than learning each method.

The other constraints are, not to collide with existing API, to be rather technical than too colourful, and to be concise and easy to memorise. `repeat` runs into the problem that you don't know whether it applies to element or sequence, and it collides with existing API:

```
a = (0..3).iter.repeat(2);
a.nextN(9)  // [0, 1, 2, 3, 0, 1, 2, 3, nil] -- therefore works at entire-sequence level
```

Also it is not a very common operation, so using a very generic term like `repeat` would be a mismatch. A good example for a method name is `shuffle`, it's immediately clear what it does (at least one has a good sense before reading up on it).

So I think `repeatEach` could work well, and more colourful `inflate` (it kind of resonates for me with the general vibe of `shuffle` etc.) I would keep looking for existing examples in other languages, even if this is not a very common operation.

Another idea, although it's not a verb, so perhaps that's less good: `redundant`.

**Comment by smrg-lm** [comment]
@Sciss sorry I changed the comment after thinking out of the phone. For me `repeatEach` is less abstract than `inflate` and more direct to the action.

**Comment by smrg-lm** [comment]
Searching for "repeat each element" gave me this:

https://stat.ethz.ch/R-manual/R-devel/library/base/html/rep.html

they use the word `replicate`. But the search engine auto-completion for "repeat each element" shows results for matlab, r, python and numpy.

**Comment by brianlheim** [comment]
fwiw, "repeatEach" was the first name i reached for intuitively when thinking about this. none of the other names suggested so far come close for me in clarity. "repeat" is a close second, but i agree with @Sciss's issues with it.

**Comment by telephon** [comment]
Just to draw together:

I see three methods:
1. use some terminology that we already use:`stretch` (good for Arrays, but already taken for Patterns), `clutch` (that exists in `Pclutch`, where it comes close to what we want), but it might have some negative connnotations which pop out especially strongly when one uses a word to get rid of another.
2. find explicit terminology (like `repeatEachBy` or similar)
3. find something that has metaphoric appeal (I like `inflate`, it hasn't got the sequential feel to it, which might be good or not. This may help: https://www.wordhippo.com/what-is/another-word-for/repeat.html


P.S./O.T.`recollect` is nice, but I'd somehow like to have a method that is more similar to `collect`: 

```
recollect { |func| var val; ^this.collect { |x, i| val = func.value(x, i) ? val } }
```

**Comment by muellmusik** [comment]
Maybe ... hold? Has good music lineage with sample and...

**Comment by Sciss** [comment]
I use `hold` for my Scala implementation of patterns to convert a single value pattern to an infinitely held pattern. So please no :)

**Comment by jamshark70** [comment]
Re: process -- if we are going to introduce backward-incompatible naming/syntax changes, we might take a page from LilyPond and ship scripts (or maybe a quark) to update user code files from one version to another. A class rename would be trivial (in Unix, a simple sed would do it).

I guess this is really for a RFC but it does strike me that, at least in this case, we can avoid the dynamic of users running an old script and getting slammed in the post window by deprecation warnings. Maybe not every case, but this one should be easy.

**Comment by NickColvin** [comment]
I'm not sure if PdurStutter was forgotten or presumed as a subclass of Pstutter, but this would naturally be replaced by Psubdivide or Psubdivision.

As for the other repetition-oriented 'stutters', 'replicate' and 'repeatEach' make intuitive sense for different cases.

Pstutter -> Preplicate - operate on the input by replicating it X times before continuing to the next input.
.stutter -> .repeatEach - repeat each element of the array/list X times.

And finally what made me chime in in the first place:

PdurStutter -> Psubdivide - subdivide the duration by x. This is far more intuitive to me as a drummer than 'PdurStutter', which is a pretty tortured metaphor and never made sense to me.

**Comment by telephon** [comment]
`Psubdivide` – yes, a good word, one should then entirely remove any reference to `dur` in the implementation.
`Preplicate` vs. `repeatEach`. Why not `replicateEach`?

**Comment by Spacechild1** [comment]
I'm stuttering and I find *zero* offence in `Pstutter`.... :shrug:. I think it's a fitting metaphor and it's not used in a negative way (except for people who think that repetition in music is bad :-p)

**Comment by brianlheim** [comment]
Hey @Spacechild1 , I'm glad that you don't find it offensive! However, that doesn't stop other people from feeling differently and asking for action to be taken about that. Our goal is to have a project that's welcoming for everyone, and unfortunately this name causes pain and distraction for some.

**Comment by Spacechild1** [comment]
> However, that doesn't stop other people from feeling differently and asking for action to be taken about that

I know, just wanted to provide a data point.

> and unfortunately this name causes pain and distraction for some.

Have you heard of *stuttering* people who feel that way? I'm really curious, because as I've said, it's one of the rare instances where the "stutter" metaphor is used in a neutral, if not productive way. I'll certainly keep using `stutter` :-)

**Comment by Sciss** [comment]
I have a lot of people with “disabilities” in my social network who are allergic to ableist language, so I think proactively avoiding these terms is good.

**Comment by timblechmann** [comment]
newspeak alert: just because something is called differently doesn't mean that people with speech disorder (or ethnicity or gender identity or body shape or social skills or age) won't be treated better.

----

that said, what about explaining the history of the "stutter" effect in the documentation (e.g. https://en.wikipedia.org/wiki/Stutter_edit), possibly with a reference of the speech disorder (https://en.wikipedia.org/wiki/Stuttering) and possibly a link to some of the classic pieces of electroacoustic music that actually have a quite poetic way of addressing the speech disorder (https://en.wikipedia.org/wiki/I_Am_Sitting_in_a_Room#Full_text)

**Comment by Spacechild1** [comment]
I think one can only really speak of "ableism" if the word is used in a clearly negative context. For example, it *is* quite painful to see old movies where stuttering is used for cheap laughs or to depict simpletons. IMO, this is not the case in `Pstutter` - on the contrary: what is usually seen as a disorder is turned into something productive. I would really love if we could approach this whole topic of "inclusive language" with a bit more nuance...

**Comment by timblechmann** [comment]
> Are you saying renaming things doesn't guarantee that people are treated differently?

renaming tends to be more greenwashing than it has *any* impact on real life

**Comment by joshpar** [comment]
Replacing language like this has become important to the SC community and it is no longer a question of removing language that may be problematic. I don't think the point of this thread is to discuss whether or not to replace an instance of something that we know some will find offensive, but to actually decide on what that replacement should be.

**Comment by NickColvin** [comment]
> `Psubdivide` – yes, a good word, one should then entirely remove any reference to `dur` in the implementation.
> `Preplicate` vs. `repeatEach`. Why not `replicateEach`?

Sure, I'd be supportive of 'replicateEach' for the list method, but there was a suggestion that 'repeatEach' would be familiar to folks coming from other languages/intuitive as a method so I was going with that.

I'm happy with either.

Only other comment I'll make to the 'stutter' discussion is that 'stutter' is not a good metaphor for anything but perhaps Pstutter, IMO, so this improves language clarity/readability as well.

**Comment by brianlheim** [comment]
> Sure, I'd be supportive of 'replicateEach' for the list method, but there was a suggestion that 'repeatEach' would be familiar to folks coming from other languages/intuitive as a method so I was going with that.

i tend to agree with this too -- this seems to be the term most people reach for based on the conversation about. do you have any issue with using this as the new name @telephon ? i'm not sure if your suggestion of `replicateEach` was meant to say you prefer it to `replicate` and `repeatEach`, or merely to offer a third alternative.

> Only other comment I'll make to the 'stutter' discussion is that 'stutter' is not a good metaphor for anything but perhaps Pstutter, IMO, so this improves language clarity/readability as well.

i totally agree with this.

i have a strong preference that whatever name is chosen, it should be used for the demand-rate ugen, pattern, and array operation. i think this is useful because it not only reinforces conceptually that the array operation is being lifted into the domain of demand ugen or pattern, but it also aids in making the pattern/ugen/array-method discoverable -- the patterns already out there lead users to expect that the pattern version of some other method is the same name with a P-prefix, and vice versa.

this would mean:

- PdurStutter -> Psubdivide
- Pstutter -> PrepeatEach
- stutter -> repeatEach
- Dstutter -> DrepeatEach

are there any concerns about this? i would suggest that we agree on something concrete soon, to keep this discussion from stagnating.

@Sciss earlier brought up the point that the `Each` suffix might not be desirable because it's uncommon in SC. i guess i would just have to say i still think it's a good name -- it seems to make sense to most people, and i haven't been drawn to most of the other suggestions without `Each`. but i also think it's good to note that we don't have precedence for this and shouldn't create too much entropy in our naming. :)

**Comment by Sciss** [comment]
> earlier brought up the point that the `Each` suffix might not be desirable because it's uncommon in SC

Just saying, unlike other language that use it, e.g. `forEach` (instead of SC `do`), there is no other use of this style yet. Not against it, I agree with the last choices.

**Comment by NickColvin** [comment]
If we are going to align the pattern/demand classes and array methods, then I would recommend going to 'replicate' for all.

My reasoning is three fold:

- repeat is used extensively throughout the language to refer to pattern/phrase repetition, whereas replicate is not
- replicate is 9 characters whereas repeatEach is 10. Also, autocomplete should offer up a pretty short list after just 4 characters 'repl'.
- repeatEach is not idiomatic

So my counter proposal is:

PdurStutter -> Psubdivide
Pstutter -> Preplicate
.stutter -> .replicate
Dstutter -> Dreplicate

**Comment by brianlheim** [comment]
To me, the single word "replicate" seems like it would cause confusion. Replicate is a synonym for duplicate. Beyond the context of this conversation, how do we justify to users that duplicate (dup) repeats arrays, while replicate repeats array elements? I agree with @Sciss 's earlier comment, "Somehow the operation *on each element* must be conveyed." Using a synonym of duplicate doesn't do this.

> repeat is used extensively throughout the language to refer to pattern/phrase repetition, whereas replicate is not

I don't understand this; we are talking about the phrase "repeat each", but you are comparing "repeat" and "replicate" here.

> repeatEach is not idiomatic

I think this is a misinterpretation of the preceding conversation. The fact that we don't have a pattern of names ending in "Each" does not make this name unidiomatic. It just means there isn't a pattern for it; it can still be a good name.

Also I want to point out that the R function noted above, which is the first mention of "replicate" in this conversation, does not do what we want our method to do. It lets you either repeat each element in the array a constant number of times or supply a second array which indicates how many times to repeat each element, AND optionally repeat the entire result a constant number of times. So their description, "replicate the values in x", is really intended to encompass the entire meaning "take the values in x and make more of them in some arrangement." Farther down, they use the phrase "repeat each element" to describe what we want our method to do.

**Comment by telephon** [comment]
Yes, exactly. The difference between `dup` and the current `stutter` is one of the level at which they operate: the first operates on the whole, the second on each item of some multiplicity (collection or stream etc.). What would be good to have is a word-particle that expresses this. We may have other cases where this will apply in the future, so this name change actually may have an extra benefit.

While literally, it would be `collect` (i.e. `x.collect { |each| each.dup(n) }`), I don't find a name like `collectDup` or `dupCollect` good, and also `mapDup` or `dupMap` are by far less clear than `stutter`. 
Good candidates are:
- `each` => `dupEach`
- `across`, `along`, or `over` => `dupOver`
- the letter `1` has sometimes be used in a similar way => `dup1`

By conclusion, something like `dupEach` would be idiomatic and systematic, if we adopt this word particle as a naming pattern. In the case of stream or ugen classes that anyhow operate on that level, we may leave it out as redundant.

```
Pstutter -> Pdup
.stutter -> .dupEach
Dstutter -> Ddup
PdurStutter -> PdurDiv // this has nothing to do with duplication
```

**Comment by brianlheim** [comment]
dup is less clear than repeat, but does have the benefit of analogy with the function called dup, so I am also fine with that as the name. I still think that the pattern and ugen should keep Each for the reasons I gave above, unless there are other situations where we do this with patterns? Otherwise, I do think it will lead to confusion and isn't worth the terseness.

I don't understand why you suggest PdurDiv now instead of Psubdivide @telephon ; that hasn't been mentioned so far as a possible replacement.

**Comment by brianlheim** [comment]
> The "each" is implied in almost all patterns and ugens, this is why I find that it doesn't really help, but may even confuse. But I don't want to force this point.

OK, thanks for explaining; I'm fine with either, I trust your intuition on that.

**Comment by NickColvin** [comment]
I am 100% on board with these options.

**Comment by telephon** [comment]
Just one thing: please in this case, let's not deprecate the old ones, but just advise users to use the new ones. Two reasons:

1. It breaks *a lot* of existing musical works
2. Technology should enable ethical behaviour, but needs (and probably should) not enforce it

**Comment by dyfer** [comment]
I believe this was solved in #5470
